:bug: Update Machine Status with Error Reason and Error Message if infra object is deleted

kashifest opened the pull request:
What this PR does / why we need it:
Updates the machine status with error reason and error message when machine is in running phase and the infrastructure object is deleted
Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
Fixes kubernetes-sigs#1398

kashifest commented:
/assign @ncdc

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+				// Infra object went missing after the machine was up and running
+				klog.Infof("Infrastructure object %v %q went missing. Updating Machine Status", ref.GroupVersionKind(), ref.Name)
```
We don't want to add any new raw `klog` calls. Please use `r.Log` instead.

kashifest replied:
ok

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 	obj, err := external.Get(r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
```
We should not check the phase - that is for human consumption. Checking InfrastructureReady is sufficient.

If we place this logic here, it applies to all external references - both the bootstrap ref and the infrastructure ref. @vincepri do you this is appropriate?

kashifest replied:
I added the machine phase to make sure this block of code is executed only once.

ncdc replied:
If we are not currently short-circuiting machine reconciliation if the error reason/error message are set, we need to be doing that.

kashifest replied:
@ncdc I am not sure if I understood correctly, this check was allowing to setting the error reason/message and logging only once, but it did not short-circuit reconciliation

ncdc replied:
In what situation would it be executing this more than once?

kashifest replied:
In each call to reconcileExternal in my current version of the code?

ncdc replied:
Ok, once you move this to reconcileInfrastructure, I think we also need to add code toward the top of Reconcile that checks machine.status.errorReason/errorMessage, and if either is set, it stops. This would need to be before the call to patch.NewHelper. Do you agree @vincepri?

vincepri replied:
Yes, the short-circuit behavior is definitely missing. Let's follow-up with a different PR marked with ⚠️ for that

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 	if err != nil {
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+				// Infra object went missing after the machine was up and running
```
The resource that went missing could be either bootstrap or infra

kashifest replied:
I thought the error status should only be set when infra resource is missing. A missing bootstrap object should not result in an error, isn't it?

detiber replied:
Bootstrap.ConfigRef is optional, someone could explicitly supply Bootstrap.Data instead.

ncdc replied:
I was referring to the case where it goes

1. Have bootstrap + infra, machine boots & is infra-ready
1. Delete bootstrap

detiber replied:
@ncdc even in that case as long as `Bootstrap.Data` is already set, do we actually care?

kashifest replied:
@ncdc I am also thinking the same as @detiber

ncdc replied:
There was someone in the community who mentioned they wanted to react to KubeadmConfig changes in their environment, so in that case, it would matter

detiber replied:
@ncdc possibly, but I wouldn't consider that a "terminal error"

ncdc replied:
Fair

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+				// Infra object went missing after the machine was up and running
+				klog.Infof("Infrastructure object %v %q went missing. Updating Machine Status", ref.GroupVersionKind(), ref.Name)
```
Could be either bootstrap or infra

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+				// Infra object went missing after the machine was up and running
+				klog.Infof("Infrastructure object %v %q went missing. Updating Machine Status", ref.GroupVersionKind(), ref.Name)
+				machineStatusError := capierrors.MachineStatusError("Invalid Configuration")
```
@vincepri do we have any guidance on the content and format for error reasons?

vincepri replied:
We should avoid creating new ErrorReason(s) dynamically. We can add new ones to the constants defined in the package. Let's focus on having a good error message in `ErrorMessage` though

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
+				klog.Infof("Infrastructure object %v %q went missing. Updating Machine Status", ref.GroupVersionKind(), ref.Name)
+				machineStatusError := capierrors.MachineStatusError("Invalid Configuration")
+				m.Status.ErrorReason = &machineStatusError
+				m.Status.ErrorMessage = pointer.StringPtr("Invalid Configuration")
```
This should be descriptive and include the GVK (GroupVersionKind) and name of the resource that went missing.

kashifest replied:
ok. For both ```ErrorReason``` and ```ErrorMessage``` ??

detiber commented on controllers/machine_controller_phases.go:
```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 	obj, err := external.Get(r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
```
We shouldn't have logic comparing against the Machine "Phase", it is meant to only be a user friendly display value.

This check should also probably be pushed up to where reconcileExternal is called for the InfrastructureRef, since a missing Bootstrap.ConfigRef should not result in an error.

kashifest replied:
I added the machine phase to make sure this block of code is executed only once.

kashifest commented:
@ncdc  I can add the test cases also. Now I am waiting to know where should I put the check in ```reconcileExternal``` or ```reoncileInfrastructure```

ncdc commented:
+1 to `reconcileInfrastructure`

vincepri commented:
Before merging, let's find a PR title that communicates better the changes we're making. PR titles go in release notes for users to read

vincepri commented:
/ok-to-test

kashifest commented:
@ncdc @vincepri @detiber I have addressed most of the issues commented above. I did not move the code to ```reconcileInfrastructure``` though, since ```reconcileExternal``` changes the error cause for reconciliation. This makes it a bit tricky to get the actual error in   ```reconcileInfrastructure``.

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +197,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
 	if infraConfig == nil && err == nil {
 		return nil
 	} else if err != nil {
+		if m.Status.InfrastructureReady {
```
Shouldn't we also check here if the error is not found?

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +197,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
+		if m.Status.InfrastructureReady {
+			// Infra or bootstrap object went missing after the machine was up and running
+			r.Log.Error(err, "Infrastructure or bootstrap object went missing. Updating Machine Status")
+			machineStatusError := capierrors.MachineStatusError("InvalidConfiguration")
```
This value should go into the `capierrors` package

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +197,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
+			r.Log.Error(err, "Infrastructure or bootstrap object went missing. Updating Machine Status")
+			machineStatusError := capierrors.MachineStatusError("InvalidConfiguration")
+			m.Status.ErrorReason = &machineStatusError
+			m.Status.ErrorMessage = pointer.StringPtr(fmt.Sprintf("Failure detected from referenced resource %v with name %q",
```
What kind of failure? Let's add a little more color on why we're setting this value

detiber commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +197,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
 	if infraConfig == nil && err == nil {
 		return nil
 	} else if err != nil {
+		if m.Status.InfrastructureReady {
```
This should probably also test `apierrors.IsNotFound(errors.Cause(err))`, to ensure that the error is related to the resource being missing rather than another type of error.

kashifest replied:
I tried that. It seems reconcileExternal is consuming that error and only returning the requeue error

detiber replied:
Arrgh, indeed it is...  I'm wondering if we need to refactor `reconcileExternal` a bit so that we are not returning a wrapped RequeueAfter error there....  It would probably require registering the watcher for the resource prior to actually retrieving one, though...  @vincepri @ncdc thoughts?

vincepri replied:
We could either move the requeue error up to the calling functions, or (less preferred, but quicker) string match "not found" on the incoming error

kashifest replied:
@vincepri @detiber I chose the quicker but not preferred way ... the placement of the requeue error in calling function is bit of a design issue since it also requires change in ```reconcileBootstrap``` , which I thought is bit out of scope of this bug, shall we follow this in another PR?

detiber replied:
I'm good with this as a short term solution as long as a tracking issue is created to clean up later.

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +197,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
 	} else if err != nil {
+		if m.Status.InfrastructureReady {
+			// Infra or bootstrap object went missing after the machine was up and running
+			r.Log.Error(err, "Infrastructure or bootstrap object went missing. Updating Machine Status")
```
```suggestion
			r.Log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
```

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -197,6 +198,14 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, m *clus
+			// Infra object went missing after the machine was up and running
+			r.Log.Error(err, "Machine infrastructure reference has been deleted after being ready, setting failure state")
+			machineStatusError := capierrors.InvalidConfigurationMachineError
+			m.Status.ErrorReason = &machineStatusError
```
```suggestion
			m.Status.ErrorReason = capierrors.ClusterStatusErrorPtr(capierrors.InvalidConfigurationMachineError)
```

kashifest replied:
> 

Didn't know this exist. Shouldn't it be MachineStatusErrorPtr instead of ClusterStatusErrorPtr

detiber commented:
/lgtm
